### PR TITLE
Userspace unit test mocks and improvements 

### PR DESCRIPTION
**Pull request by ctsims**
Implemented some mocks which should make it super easy to build and execute tests against the user space sandbox from XML restore files, and quick reference integration test for making sure that the purges are working

Relies on: https://github.com/dimagi/javarosa/pull/107

### CODE REVIEW

File: tests/test/org/commcare/util/test/CasePurgeIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    public void tearDown() throws Exception {
+    }
+
+    @Test
+    public void test() {
+        CasePurgeFilter purger = new CasePurgeFilter(sandbox.getCaseStorage(), owners);
+        int removedCases = sandbox.getCaseStorage().removeAll(purger).size();
+        
+        if(removedCases == 0) {
+           throw new RuntimeException("Failed to remove purged case");
```
Comment (phillipm) [inline tests/test/org/commcare/util/test/CasePurgeIntegrationTest.java]: Test fails here. Did I set it up wrong?

Comment (phillipm) [inline tests/test/org/commcare/util/test/CasePurgeIntegrationTest.java]: Hold on, may not have setup resources correctly

File: util/src/org/commcare/util/mocks/MockDataUtils.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        return owners;
+    }
+    
+    /**
+     * Load the referenced fixture out of storage for the provided user 
+     * 
+     * @param sandbox
+     * @param refId
+     * @param userId
+     * @return
```
Comment (phillipm) [inline util/src/org/commcare/util/mocks/MockDataUtils.java]: I prefer to not include argument param annotations unless they also have corresponding comments.

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    public TransactionParser getParser(KXmlParser parser) {
+        String namespace = parser.getNamespace();
+        String name = parser.getName();
+        if (LedgerXmlParsers.STOCK_XML_NAMESPACE.matches(namespace)) {
+            if (stockParser == null) {
+                throw new RuntimeException("Couldn't process Stock transaction without initialization!");
+            }
+            req();
+            return stockParser.getParser(parser);
+        } else if (name != null && name.toLowerCase().equals("case")) {
```
Comment (phillipm) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: how about remove the null check and just have: "case".equalsIgnoreCase(name) ?

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            //server message;
+            //" <message nature=""/>"
+        } else if (name != null && name.toLowerCase().equals("sync") && namespace != null && "http://commcarehq.org/sync".equals(namespace)) {
+            return new TransactionParser<String>(parser) {
+                /*
+                 * (non-Javadoc)
+                 * @see org.commcare.data.xml.TransactionParser#commit(java.lang.Object)
+                 */
+                @Override
+                public void commit(String parsed) throws IOException {
```
Comment (phillipm) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: I don't like the (non-Javadoc) comments when there is an explicit Override annotation. It is easy to see what is being overridden, no need for messy-looking comments.

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+            }
+            req();
+            return userParser.getParser(parser);
+        } else if (name != null && name.toLowerCase().equals("fixture")) {
+            req();
+            return fixtureParser.getParser(parser);
+        } else if (name != null && name.toLowerCase().equals("message")) {
+            //server message;
+            //" <message nature=""/>"
+        } else if (name != null && name.toLowerCase().equals("sync") && namespace != null && "http://commcarehq.org/sync".equals(namespace)) {
```
Comment (phillipm) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: No need for a "namespace != null" check

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    public void initStockParser() {
+        stockParser = new TransactionParserFactory() {
+
+            public TransactionParser<Ledger[]> getParser(KXmlParser parser) {
+                return new LedgerXmlParsers(parser, sandbox.getLedgerStorage());
+            }
+        };
+    }
+
+    public String getSyncToken() {
```
Comment (phillipm) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: This is never called, probably worth removing

File: util/src/org/commcare/util/mocks/MockDataUtils.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * Load the referenced fixture out of storage for the provided user
+     *
+     * @param sandbox
+     * @param refId
+     * @param userId
+     * @return
+     */
+    private static FormInstance loadFixture(MockUserDataSandbox sandbox,
```
Comment (phillipm) [inline util/src/org/commcare/util/mocks/MockDataUtils.java]: I prefer to not include argument param annotations unless they also have corresponding comments.

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            return userParser.getParser(parser);
+        } else if (name != null && name.toLowerCase().equals("fixture")) {
+            req();
+            return fixtureParser.getParser(parser);
+        } else if (name != null && name.toLowerCase().equals("message")) {
+            //server message;
+            //" <message nature=""/>"
+        } else if (name != null && name.toLowerCase().equals("sync") && 
+                namespace != null && "http://commcarehq.org/sync".equals(namespace)) {
+            return new TransactionParser<String>(parser) {
```
Comment (phillipm) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: No need for a "namespace != null" check

**Comment by phillipm** [comment]
Hooray for more tests!

@ctsims Done reviewing. Will merge after you address my minor comments. Hope you don't mind my neurotic lint and style commits.

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                            return fixtureStorage;
+                        }
+                    };
+                }
+
+                return created;
+            }
+        };
+
+        this.initUserParser();
```
Comment (wpride) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: is there a reason the fixture parser isn't in a initFixtureParser() function like these others?

**Comment by wpride** [comment]
Is there/should there be a readme included with this?

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java (hunk 9 above)
Comment (ctsims) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: Just legacy (IE: As mentioned in comments this was copied from Android, so it hadn't gotten moved yet).

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java (hunk 3 above)
Comment (ctsims) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: Agree with this pattern in general. Fixing

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java (hunk 4 above)
Comment (ctsims) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: I believe those are there so automated javadoccer's can redirect to the parent page, but we should verify.

File: util/src/org/commcare/util/CommCareTransactionParserFactory.java (hunk 6 above)
Comment (ctsims) [inline util/src/org/commcare/util/CommCareTransactionParserFactory.java]: adding this to the integration test

**Comment by ctsims** [comment]
@phillipm @wspride 

Updated re:reviewing. 

Will - A readme is a good idea for the harness stuff at commandline for sure. I'll prioritize that for the command line util stuff, but don't think it affects the tests as much.

**Comment by wpride** [comment]
@ctsims I've been using this and its all awesome

**Pull request merged**